File: benchmarks/livermore_canonical/k11_prefix_sum.f
C     Kernel 11 -- First sum / partial sums (canonical)
C     From "The Livermore Fortran Kernels" (UCRL-53745), F.H. McMahon, 1986
C     Canonical body: X(1) = Y(1); DO k=2,n: X(k) = X(k-1) + Y(k)
C
      PROGRAM K11
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      INTEGER REP
      DIMENSION X(1001), Y(1001)
C
C     Initialize Y
C
      FUZZ  = 1.2345D-3
      BUZZ  = 1.0D0 + FUZZ
      FIZZ  = 1.1D0 * FUZZ
      DO 10 K = 1, 1001
          BUZZ = (1.0D0 - FUZZ) * BUZZ + FUZZ
          FUZZ = -FUZZ
          Y(K) = (BUZZ - FIZZ) * 0.1D0
   10 CONTINUE
C
      N = 1001
      DO 200 REP = 1, 34000000
          X(1) = Y(1)
          DO 90 K = 2, N
              X(K) = X(K-1) + Y(K)
   90     CONTINUE
  200 CONTINUE
C
      WRITE(*,900) X(N)
  900 FORMAT('x(n) = ', E25.15)
      END

File: benchmarks/livermore_canonical/k11_prefix_sum.c
/* K11 -- First sum / partial sums (canonical, 1-based indexing)
   Canonical body: X(1) = Y(1); DO k=2,n: X(k) = X(k-1) + Y(k) */
#include <stdio.h>
#include <time.h>
#include "signel.h"

#define N     1001
#define NREPS 30000000

int main(void) {
    double x[1002], y[1002];

    signel(y+1, 1001);

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    for (long rep = 0; rep < NREPS; rep++) {
        x[1] = y[1];
        for (long k = 2; k <= N; k++) {
            x[k] = x[k-1] + y[k];
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) * 1e-9;
    printf("elapsed: %.6f s\n", elapsed);
    printf("x[n] = %f\n", x[N]);
    return 0;
}
